src/object/srv_obj_migrate.c

```diff
@@ -475,6 +475,57 @@ migrate_pool_tls_lookup_create(struct ds_pool *pool, int version,
 	return tls;
 }
 
+static void
+mrone_recx_daos_vos_internal(struct migrate_one *mrone,
+			     struct daos_oclass_attr *oca,
+			     bool daos2vos, int shard)
+{
+	daos_iod_t *iod;
+	int cell_nr;
+	int stripe_nr;
+	int j;
+	int k;
+
+	D_ASSERT(DAOS_OC_IS_EC(oca));
+
+	cell_nr = obj_ec_cell_rec_nr(oca);
+	stripe_nr = obj_ec_stripe_rec_nr(oca);
+	/* Convert the DAOS to VOS EC offset */
+	for (j = 0; j < mrone->mo_iod_num; j++) {
+		iod = &mrone->mo_iods[j];
+		for (k = 0; k < iod->iod_nr; k++) {
+			daos_recx_t *recx;
+
+			recx = &iod->iod_recxs[k];
+			D_ASSERT(recx->rx_nr <= cell_nr);
+			if (daos2vos)
+				recx->rx_idx = obj_ec_idx_daos2vos(recx->rx_idx,
+								   stripe_nr,
+								   cell_nr);
+			else
+				recx->rx_idx = obj_ec_idx_vos2daos(recx->rx_idx,
+								   stripe_nr,
+								   cell_nr,
+								   shard);
+			D_DEBUG(DB_REBUILD, "j %d k %d "DF_U64"/"DF_U64"\n",
+				j, k, recx->rx_idx, recx->rx_nr);
+		}
+	}
+}
+
+static void
+mrone_recx_daos2_vos(struct migrate_one *mrone, struct daos_oclass_attr *oca)
+{
+	mrone_recx_daos_vos_internal(mrone, oca, true, -1);
+}
+
+static void
+mrone_recx_vos2_daos(struct migrate_one *mrone, struct daos_oclass_attr *oca,
+		     int shard)
+{
+	mrone_recx_daos_vos_internal(mrone, oca, false, shard);
```

Comment:
for vos2_daos, the obj_ec_idx_vos2daos should pass shard_idx rather than direct shard value, so seems should either translate to shard_idx here or before calling obj_ec_idx_vos2daos().

Reply:
oh, yes, you are right.